File: golang/xidl-go-rest/websocket.go
package xidlgohttp

import (
	"context"
	"errors"
	"io"
	"net"
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
)

const defaultWSWriteWait = 10 * time.Second

// WSBidiStream is a full-duplex typed WebSocket session (RFC 6455).
// One application message equals one WebSocket Text frame (JSON codec).
// TRead is the decoded inbound application message type.
// TWrite is the outbound application message type.
type WSBidiStream[TRead any, TWrite any] struct {
	conn *websocket.Conn
}

// NewWSBidiServer wraps an upgraded connection as a server-side session.
func NewWSBidiServer[TIn any, TOut any](conn *websocket.Conn) *WSBidiStream[TIn, TOut] {
	return &WSBidiStream[TIn, TOut]{conn: conn}
}

// NewWSBidiClient wraps a dialed connection as a client-side session.
// TIn is what the client writes, TOut is what the client reads.
func NewWSBidiClient[TIn any, TOut any](conn *websocket.Conn) *WSBidiStream[TOut, TIn] {
	return &WSBidiStream[TOut, TIn]{conn: conn}
}

// Read returns the next inbound application message.
// It returns io.EOF when the peer closes the stream cleanly.
func (s *WSBidiStream[TRead, TWrite]) Read() (TRead, error) {
	var zero TRead
	for {
		msgType, data, err := s.conn.ReadMessage()
		if err != nil {
			if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) ||
				errors.Is(err, io.EOF) || errors.Is(err, net.ErrClosed) ||
				websocket.IsUnexpectedCloseError(err) {
				return zero, io.EOF
			}
			return zero, err
		}
		switch msgType {
		case websocket.TextMessage, websocket.BinaryMessage:
			var value TRead
			if err := MustCodecForMime("application/json").Decode(strings.NewReader(string(data)), &value); err != nil {
				return zero, &HttpError{
					Status:  http.StatusBadRequest,
					Code:    http.StatusBadRequest,
					Message: "invalid ws payload: " + err.Error(),
				}
			}
			return value, nil
		case websocket.PingMessage:
			_ = s.conn.WriteControl(websocket.PongMessage, data, time.Now().Add(defaultWSWriteWait))
		case websocket.PongMessage, websocket.CloseMessage:
			if msgType == websocket.CloseMessage {
				return zero, io.EOF
			}
		default:
		}
	}
}

// Write sends one outbound application message as a WebSocket Text frame.
func (s *WSBidiStream[TRead, TWrite]) Write(value TWrite) error {
	var buf strings.Builder
	if err := MustCodecForMime("application/json").Encode(&buf, value); err != nil {
		return err
	}
	return s.conn.WriteMessage(websocket.TextMessage, []byte(buf.String()))
}

// Close sends a normal Close frame and releases the connection.
func (s *WSBidiStream[TRead, TWrite]) Close() error {
	msg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")
	_ = s.conn.WriteControl(websocket.CloseMessage, msg, time.Now().Add(defaultWSWriteWait))
	return s.conn.Close()
}

// UpgradeWebSocket upgrades a gin request to RFC 6455 WebSocket, optionally
// requiring a single subprotocol when subprotocols has exactly one entry.
func UpgradeWebSocket(c *gin.Context, subprotocols []string) (*websocket.Conn, error) {
	if len(subprotocols) == 1 {
		offered := websocket.Subprotocols(c.Request)
		if len(offered) > 0 && !containsString(offered, subprotocols[0]) {
			return nil, &HttpError{
				Status:  http.StatusBadRequest,
				Code:    http.StatusBadRequest,
				Message: "missing required Sec-WebSocket-Protocol",
			}
		}
	}
	upgrader := websocket.Upgrader{
		CheckOrigin:  func(r *http.Request) bool { return true },
		Subprotocols: subprotocols,
	}
	return upgrader.Upgrade(c.Writer, c.Request, nil)
}

// DialWebSocket opens a client WebSocket connection to endpoint.
func DialWebSocket(ctx context.Context, endpoint string, subprotocols []string) (*websocket.Conn, error) {
	dialer := websocket.Dialer{
		Subprotocols: subprotocols,
		Proxy:        http.ProxyFromEnvironment,
	}
	header := http.Header{}
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	conn, _, err := dialer.DialContext(ctx, endpoint, header)
	if err != nil {
		return nil, err
	}
	return conn, nil
}

func containsString(values []string, want string) bool {
	for _, v := range values {
		if strings.EqualFold(v, want) {
			return true
		}
	}
	return false
}

File: typescript/xidl-typescript-server/src/websocket.ts
import { deserialize, serialize, type XidlSchema } from 'xidl-typescript-codec';

/** Typed full-duplex session backed by a WebSocket-like connection. */
export interface WsBidiServerSession<TIn, TOut> {
  read(): Promise<TIn | null>;
  write(value: TOut): Promise<void>;
  close(code?: number, reason?: string): void;
}

export type WsLikeSocket = {
  send(data: string): void;
  close(code?: number, reason?: string): void;
  on(event: string, listener: (...args: unknown[]) => void): void;
  readyState?: number;
  protocol?: string;
};

/**
 * Wrap an accepted WebSocket connection (for example from the `ws` package)
 * as a typed XIDL bidi session. One application message equals one Text frame.
 */
export function wrapWsBidiServer<TIn, TOut>(
  socket: WsLikeSocket,
  options?: { inSchema?: XidlSchema; outSchema?: XidlSchema },
): WsBidiServerSession<TIn, TOut> {
  const queue: TIn[] = [];
  let notify: ((value: TIn | null) => void) | null = null;
  let closed = false;

  socket.on('message', (...args: unknown[]) => {
    const data = args[0];
    const raw = typeof data === 'string' ? data : String(data);
    const parsed = JSON.parse(raw);
    const value = (
      options?.inSchema ? deserialize(parsed, options.inSchema) : parsed
    ) as TIn;
    if (notify) {
      const resolve = notify;
      notify = null;
      resolve(value);
    } else {
      queue.push(value);
    }
  });
  socket.on('close', () => {
    closed = true;
    if (notify) {
      const resolve = notify;
      notify = null;
      resolve(null);
    }
  });
  socket.on('error', () => {
    closed = true;
    if (notify) {
      const resolve = notify;
      notify = null;
      resolve(null);
    }
  });

  return {
    close(code = 1000, reason = '') {
      closed = true;
      socket.close(code, reason);
    },
    async read() {
      if (queue.length > 0) {
        return queue.shift() as TIn;
      }
      if (closed) {
        return null;
      }
      return new Promise<TIn | null>(resolve => {
        notify = resolve;
      });
    },
    async write(value: TOut) {
      const payload = options?.outSchema
        ? serialize(value, options.outSchema)
        : value;
      socket.send(JSON.stringify(payload));
    },
  };
}

export interface UpgradeWebSocketOptions {
  subprotocol?: string;
}

/**
 * Validate an HTTP upgrade request for RFC 6455 and pick a subprotocol.
 * Returns false when the request is not a valid WebSocket upgrade,
 * null when no subprotocol is required, or the selected subprotocol.
 */
export function selectWsSubprotocol(
  request: Request,
  options?: UpgradeWebSocketOptions,
): string | null | false {
  const upgrade = request.headers.get('upgrade') ?? '';
  if (!upgrade.toLowerCase().includes('websocket')) {
    return false;
  }
  if (!options?.subprotocol) {
    return null;
  }
  const offered = (request.headers.get('sec-websocket-protocol') ?? '')
    .split(',')
    .map(value => value.trim())
    .filter(Boolean);
  if (
    offered.length > 0 &&
    !offered.some(value => value === options.subprotocol)
  ) {
    return false;
  }
  return options.subprotocol;
}
